packages/contractkit/src/identity/offchain/schemas/utils.ts

```diff
@@ -0,0 +1,348 @@
+import { Err, Ok, parseJsonAsResult, Result, trimLeading0x } from '@celo/base/src'
+import { publicKeyToAddress } from '@celo/utils/lib/address'
+import { AES128Decrypt, Encrypt } from '@celo/utils/lib/ecies'
+import { EIP712Object, EIP712TypedData } from '@celo/utils/lib/sign-typed-data-utils'
+import { trimPublicKeyPrefix } from '@celo/utils/src/ecdh'
+import { createCipheriv, createHmac, randomBytes } from 'crypto'
+import { keccak256 } from 'ethereumjs-util'
+import { isLeft } from 'fp-ts/lib/Either'
+import * as t from 'io-ts'
+import { join, normalize } from 'path'
+import OffchainDataWrapper, { OffchainErrorTypes } from '../../offchain-data-wrapper'
+import {
+  InvalidDataError,
+  InvalidKey,
+  OffchainError,
+  SchemaErrors,
+  SchemaErrorTypes,
+  UnavailableKey,
+} from './errors'
+
+export interface Schema<DataType> {
+  write: (data: DataType) => Promise<SchemaErrors | void>
+  read: (from: string) => Promise<DataType>
+  readAsResult: (from: string) => Promise<Result<DataType, SchemaErrors>>
+}
+
+export interface EncryptedSchema<DataType> {
+  write: (data: DataType, to: string[], symmetricKey?: Buffer) => Promise<SchemaErrors | void>
+  read: (from: string) => Promise<DataType>
+  readAsResult: (from: string) => Promise<Result<DataType, SchemaErrors>>
+}
+
+const KEY_LENGTH = 16
+const IV_LENGTH = 16
+
+// label = PRF(ECDH(A, B), A || B || data path)
+// ciphertext path = "/cosmetic path/" || base64(label)
+function getCiphertextLabel(
+  path: string,
+  sharedSecret: Buffer,
+  senderPublicKey: string,
+  receiverPublicKey: string
+) {
+  const senderPublicKeyBuffer = Buffer.from(trimLeading0x(senderPublicKey), 'hex')
+  const receiverPublicKeyBuffer = Buffer.from(trimLeading0x(receiverPublicKey), 'hex')
+
+  const label = createHmac('blake2s256', sharedSecret)
+    .update(Buffer.concat([senderPublicKeyBuffer, receiverPublicKeyBuffer, Buffer.from(path)]))
+    .digest('base64')
+  return normalize(join('ciphertexts', label))
```

Comment:
So this may be problematic because `/` is a valid base 64 character. For instance, if the label begins with `/`, or has a repeated `/` in the middle, normalize with break the path by changing `//` to `/`. It also reveals we may already have an issue, with or without normalization, in that Google Storage may interpret repeated `/` incorrectly as it needs to be treated as different than a single `/`. We may need to use and require use of RFC 4648 Section 5 alphabet https://tools.ietf.org/html/rfc4648#section-5, or think of another way to avoid this issue.

Reply:
I don't know if this concerns us.

How the google storage interface displays "folders" is irrelevant for us. The file `gs://hello/test//test` is just as readable and writeable as `gs://hello/test/test`

Reply:
Sorry I didn't make this point very clear. I think we have a potential correctness issue in the edge case that the `base64` ciphertext label either starts with a slash, or contains a repeated slash. How this potential issue manifests is largely dependent on the specific handling by the storage provider, so for now just GCS, but in the future we want to support arbitrary implementations.

I realize now that using `base64` is likely a mistake because of this sharp-edge. Instead I think we should change  line 49 here to use `hex`, and back propagate the change to the spec. (IIUC, this should be a one-line change here now, but would be a must bigger issue if we ever wanted to migrate in the future) I think the only downside to using `hex` is that the ciphertext label will be 64 characters instead of 42, but that seems trivial.